### PR TITLE
impl(pubsub): use option to determine link batch size

### PR DESCRIPTION
**Pull request by alevenberg**
Closes https://github.com/googleapis/google-cloud-cpp/issues/12979

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13047)

### CODE REVIEW

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 1]
```diff
@@ -91,12 +92,13 @@ auto MakeChild(
 }
 
 Spans MakeBatchSinkSpans(Spans message_spans) {
-  int64_t max_otel_links = 128;
+  size_t max_otel_links =
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:95]: We spell `std::size_t`, or `auto const` unless you need to silence warnings about signed/unsigned comparisons.

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc [hunk 2]
```diff
@@ -208,6 +213,8 @@ TEST(TracingMessageBatch, FlushSmallBatch) {
 TEST(TracingMessageBatch, FlushBatchWithOtelLimit) {
   namespace sc = ::opentelemetry::trace::SemanticConventions;
   auto constexpr kBatchSize = 128;
+  internal::OptionsSpan options(
+      Options{}.set<pubsub::MaxOtelLinkCountOption>(128));
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch_test.cc:217]: nit: s/128/kBatchSize/

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc [hunk 3]
```diff
@@ -236,6 +243,8 @@ TEST(TracingMessageBatch, FlushBatchWithOtelLimit) {
 
 TEST(TracingMessageBatch, FlushLargeBatch) {
   namespace sc = ::opentelemetry::trace::SemanticConventions;
+  internal::OptionsSpan options(
+      Options{}.set<pubsub::MaxOtelLinkCountOption>(128));
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch_test.cc:247]: nit: s/128/kBatchSize - 1/ ?

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 4]
```diff
@@ -91,12 +92,13 @@ auto MakeChild(
...
+      internal::CurrentOptions().get<pubsub::MaxOtelLinkCountOption>();
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:96]: Blegh... is there a way to get the options explicitly?  Or are they completely lost in the call stack by this point?

**Review by coryan — APPROVED** [review]
A few nits.  The `128` magic appears in multiple places. At least within the same file consider using a constant.

File: google/cloud/pubsub/internal/tracing_message_batch.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc]: Yeah that would be better. Fixed. Refactored so the class takes an options.

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc (hunk 3 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch_test.cc]: Done

File: google/cloud/pubsub/internal/tracing_message_batch_test.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch_test.cc]: Done

File: google/cloud/pubsub/internal/tracing_message_batch.cc [hunk 5]
```diff
@@ -151,7 +153,8 @@ class TracingMessageBatch : public MessageBatch {
       message_spans.swap(message_spans_);
     }
 
-    auto batch_sink_spans = MakeBatchSinkSpans(std::move(message_spans));
+    auto batch_sink_spans =
+        MakeBatchSinkSpans(std::move(message_spans), std::move(options_));
```
Comment (devbww) [inline google/cloud/pubsub/internal/tracing_message_batch.cc:157]: It is strange to move from a member variable (at least not within an rvalue ref-qualified function).

Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_message_batch.cc]: Ack, done.

**Pull request merged**